src/algebra/modeq.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.int.modeq
+import group_theory.quotient_group
+
+/-!
+# Equality modulo an element
+
+This file defines equality modulo an element in a commutative group.
+
+## Main definitions
+
+* `a ≡ b [PMOD p]`: `a` and `b` are congruent modulo a`p`.
+
+## See also
+
+`smodeq` is a generalisation to arbitrary submodules.
+
+## TODO
+
+Delete `int.modeq` in favour of `add_comm_group.modeq`. Generalise `smodeq` to `add_subgroup` and
+redefine `add_comm_group.modeq` using it. Once this is done, we can rename `add_comm_group.modeq`
+to `add_subgroup.modeq` and multiplicativise it. Longer term, we could generalise to submonoids and
+also unify with `nat.modeq`.
+-/
+
+namespace add_comm_group
+variables {α : Type*}
+
+section add_comm_group
+variables [add_comm_group α] {p a a₁ a₂ b b₁ b₂ c : α} {n : ℕ} {z : ℤ}
+
+/-- `a ≡ b [PMOD p]` means that `b` is congruent to `a` modulo `p`.
+
+Equivalently (as shown in `algebra.order.to_interval_mod`), `b` does not lie in the open interval
+`(a, a + p)` modulo `p`, or `to_Ico_mod hp a` disagrees with `to_Ioc_mod hp a` at `b`, or
+`to_Ico_div hp a` disagrees with `to_Ioc_div hp a` at `b`. -/
+def modeq (p a b : α) : Prop := ∃ z : ℤ, b - a = z • p
+
+notation a ` ≡ `:50 b ` [PMOD `:50 p `]`:0 := modeq p a b
+
+@[refl, simp] lemma modeq_refl (a : α) : a ≡ a [PMOD p] := ⟨0, by simp⟩
+
+lemma modeq_rfl : a ≡ a [PMOD p] := modeq_refl _
+
+lemma modeq_comm : a ≡ b [PMOD p] ↔ b ≡ a [PMOD p] :=
+(equiv.neg _).exists_congr_left.trans $ by simp [modeq, ←neg_eq_iff_eq_neg]
+
+alias modeq_comm ↔ modeq.symm _
+
+attribute [symm] modeq.symm
+
+@[trans] lemma modeq.trans : a ≡ b [PMOD p] → b ≡ c [PMOD p] → a ≡ c [PMOD p] :=
+λ ⟨m, hm⟩ ⟨n, hn⟩, ⟨m + n, by simp [add_smul, ←hm, ←hn]⟩
+
+instance : is_refl _ (modeq p) := ⟨modeq_refl⟩
+
+@[simp] lemma neg_modeq_neg : -a ≡ -b [PMOD p] ↔ a ≡ b [PMOD p] :=
+modeq_comm.trans $ by simp [modeq]
+
+alias neg_modeq_neg ↔ modeq.of_neg modeq.neg
+
+@[simp] lemma modeq_neg : a ≡ b [PMOD -p] ↔ a ≡ b [PMOD p] :=
+modeq_comm.trans $ by simp [modeq, ←neg_eq_iff_eq_neg]
+
+alias modeq_neg ↔ modeq.of_neg' modeq.neg'
+
+@[simp] lemma zsmul_modeq_zsmul [no_zero_smul_divisors ℤ α] (hn : z ≠ 0) :
+  z • a ≡ z • b [PMOD (z • p)] ↔ a ≡ b [PMOD p] :=
+exists_congr $ λ m, by rw [←smul_sub, smul_comm, smul_right_inj hn]; apply_instance
+
+@[simp] lemma nsmul_modeq_nsmul [no_zero_smul_divisors ℕ α] (hn : n ≠ 0) :
+  n • a ≡ n • b [PMOD (n • p)] ↔ a ≡ b [PMOD p] :=
+exists_congr $ λ m, by rw [←smul_sub, smul_comm, smul_right_inj hn]; apply_instance
+
+alias zsmul_modeq_zsmul ↔ modeq.zsmul_cancel _
+alias nsmul_modeq_nsmul ↔ modeq.nsmul_cancel _
```

Comment:
It's a bit confusing having these stronger results before `zsmul` below. Can you move this down to be next to them?